File: problem2.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// using a HashMap to store the count of the cumulative difference between
// the number of zeros and ones encountered while traversing the array

import java.util.HashMap;
import java.util.Map;

public class problem2 {

    public int findMaxLength(int[] nums) {
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, -1);
        int maxlen = 0, count = 0;
        for (int i = 0; i < nums.length; i++) {
            count = count + (nums[i] == 1 ? 1 : -1);
            if (map.containsKey(count)) {
                maxlen = Math.max(maxlen, i - map.get(count));
            } else {
                map.put(count, i);
            }
        }
        return maxlen;
    }
}

File: Problem-2.ts
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes 
// Any problem you faced while coding this : it's failing when I am trying to submit in leet code
function findMaxLength(nums: number[]): number {
    let map = { 0: -1 };
    let count = 0;
    let maxLength = 0;

    for (let i = 0; i < nums.length; i++) {
        // if 0 decrease the count, or if 1 increase the count
        count = count + (nums[i] === 0 ? -1 : 1);
        // if count exists in the map then get the max
        if (map[count]) {
            maxLength = Math.max(maxLength, i - map[count]);
        } else if(count === 0){
            maxLength = i + 1;
        }
         else {
            // set the map to i 
            map[count] = i;
        }
    }
    return maxLength;
};